

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>User guide &mdash; sinto 0.7.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Processing scATAC-seq data" href="scatac.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> sinto
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">User guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#create-scatac-seq-fragments-file">Create scATAC-seq fragments file</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fragment-file-format">Fragment file format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#how-the-fragment-file-is-generated">How the fragment file is generated</a></li>
<li class="toctree-l3"><a class="reference internal" href="#additional-arguments-for-the-fragments-function">Additional arguments for the fragments function</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#number-of-processors-nproc">Number of processors: <code class="docutils literal notranslate"><span class="pre">--nproc</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#minimum-mapping-quality-min-mapq">Minimum mapping quality: <code class="docutils literal notranslate"><span class="pre">--min_mapq</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#cell-barcode-tag-barcodetag">Cell barcode tag: <code class="docutils literal notranslate"><span class="pre">--barcodetag</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#cell-barcode-regex-barcode-regex">Cell barcode regex: <code class="docutils literal notranslate"><span class="pre">--barcode_regex</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#choosing-chromosomes-to-include-use-chrom">Choosing chromosomes to include: <code class="docutils literal notranslate"><span class="pre">--use_chrom</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#set-the-maximum-distance-between-tn5-integration-sites-max-distance">Set the maximum distance between Tn5 integration sites: <code class="docutils literal notranslate"><span class="pre">--max_distance</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#set-the-maximum-number-of-fragments-to-hold-in-memory-before-collapsing-chunksize">Set the maximum number of fragments to hold in memory before collapsing: <code class="docutils literal notranslate"><span class="pre">--chunksize</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#filter-cell-barcodes-from-bam-file">Filter cell barcodes from BAM file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#add-read-tags-to-bam-file">Add read tags to BAM file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#convert-read-tag-to-read-group">Convert read tag to read group</a></li>
<li class="toctree-l2"><a class="reference internal" href="#copy-move-read-tag-to-another-read-tag">Copy/move read tag to another read tag</a></li>
<li class="toctree-l2"><a class="reference internal" href="#add-cell-barcodes-to-fastq-read-names">Add cell barcodes to FASTQ read names</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#additional-arguments-for-the-barcode-function">Additional arguments for the barcode function</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#bases-bases">Bases: <code class="docutils literal notranslate"><span class="pre">--bases</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#barcode-read-file-barcode-fastq">Barcode read file: <code class="docutils literal notranslate"><span class="pre">--barcode_fastq</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#read-1-and-read-2-read1-and-read2">Read 1 and read 2: <code class="docutils literal notranslate"><span class="pre">--read1</span></code> and <code class="docutils literal notranslate"><span class="pre">--read2</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="scatac.html">Processing scATAC-seq data</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sinto</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>User guide</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/basic_usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="user-guide">
<h1>User guide<a class="headerlink" href="#user-guide" title="Permalink to this headline">¶</a></h1>
<div class="section" id="create-scatac-seq-fragments-file">
<h2>Create scATAC-seq fragments file<a class="headerlink" href="#create-scatac-seq-fragments-file" title="Permalink to this headline">¶</a></h2>
<p>An ATAC-seq fragment file can be created from a BAM file using the <code class="docutils literal notranslate"><span class="pre">fragments</span></code> command.
The fragment file contains the position of each Tn5 integration site, the cell barcode
associated with the fragment, and the number of times the fragment was sequenced.
PCR duplicates are collapsed.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> sinto fragments [-h] -b BAM -f FRAGMENTS [-m MIN_MAPQ] [-p NPROC]
                    [-t BARCODETAG] [-c CELLS]
                    [--barcode_regex BARCODE_REGEX] [--use_chrom USE_CHROM]
                    [--max_distance MAX_DISTANCE]

 Create ATAC-seq fragment file from BAM file

 optional arguments:
 -h, --help            show this help message and exit
 -b BAM, --bam BAM     Input bam file (must be indexed)
 -f FRAGMENTS, --fragments FRAGMENTS
                         Name and path for output fragments file. Note that the
                         output is not sorted or compressed. To sort the output
                         file use sort -k 1,1 -k2,2n
 -m MIN_MAPQ, --min_mapq MIN_MAPQ
                         Minimum MAPQ required to retain fragment (default =
                         30)
 -p NPROC, --nproc NPROC
                         Number of processors (default = 1)
 -t BARCODETAG, --barcodetag BARCODETAG
                         Read tag storing cell barcode information (default =
                         &quot;CB&quot;)
 -c CELLS, --cells CELLS
                         Path to file containing cell barcodes to retain, or a
                         comma-separated list of cell barcodes. If None
                         (default), use all cell barocodes present in the BAM
                         file.
 --barcode_regex BARCODE_REGEX
                         Regular expression used to extract cell barcode from
                         read name. If None (default), extract cell barcode
                         from read tag. Use &quot;[^:]*&quot; to match all characters up
                         to the first colon.
 --use_chrom USE_CHROM
                         Regular expression used to match chromosomes to be
                         included in output. Default is &quot;(?i)^chr&quot; to match all
                         chromosomes starting with &quot;chr&quot;, case insensitive
 --max_distance MAX_DISTANCE
                         Maximum distance between integration sites for the
                         fragment to be retained. Allows filtering of
                         implausible fragments that likely result from
                         incorrect mapping positions. Default is 5000 bp.
--chunksize CHUNKSIZE
                        Number of BAM file entries to iterate over before
                        collapsing the fragments and writing to disk. Higher
                        chunksize will use more memory but will be faster.
</pre></div>
</div>
<div class="section" id="fragment-file-format">
<h3>Fragment file format<a class="headerlink" href="#fragment-file-format" title="Permalink to this headline">¶</a></h3>
<p>The fragment file is a BED format file containing the positions of Tn5 integration
sites, the cell barcode that the DNA fragment originated from, and the number
of times the fragment was sequenced. See the <a class="reference external" href="https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/fragments">10x Genomics</a>
website for a further description of the fragment file format.</p>
<p>This is a convenient compressed form of the most useful data generated in a
scATAC-seq experiment. The fragments file generated by Sinto needs to be sorted,
block-gzip compressed (<a class="reference external" href="https://www.htslib.org/doc/bgzip.html">bgzip</a>),
and indexed using <a class="reference external" href="https://www.htslib.org/doc/tabix.html">tabix</a>.</p>
<p>For example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sort -k <span class="m">1</span>,1 -k2,2n frags.bed &gt; frags.sort.bed
bgzip frags.sort.bed
tabix -p bed frags.sort.bed
</pre></div>
</div>
</div>
<div class="section" id="how-the-fragment-file-is-generated">
<h3>How the fragment file is generated<a class="headerlink" href="#how-the-fragment-file-is-generated" title="Permalink to this headline">¶</a></h3>
<p>Generating the fragment file involves the following steps in order:</p>
<ol class="arabic">
<li><p>Remove soft-clipped bases from the alignment position.</p></li>
<li><p>Extract cell barcode sequence associated with the fragment.</p></li>
<li><p>Adjust alignment positions for the 9 bp Tn5 shift by
applying +4/-5 to the start and end position of the paired reads.</p></li>
<li><p>Remove fragments where either read has a MAPQ score less than
the specified cutoff.</p></li>
<li><p>Remove fragments where the fragment size is greater than the
specified maximum.</p></li>
<li><p>Collapse PCR duplicates:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Count the frequency of each fragment for each cell barcode.</p></li>
<li><p>Within a cell barocode, collapse fragments that share
a start or end coordinate on the same chromosome.</p></li>
<li><p>Across all cell barcodes, collapse fragments that share
the exact start and end coordinates on the same chromosome.</p></li>
<li><p>Assign the fragment to the most abundant cell barcode.</p></li>
<li><p>Record the read count for the collapsed fragment.</p></li>
</ol>
</div></blockquote>
</li>
<li><p>Write fragments to file. Note that fragments are not sorted
or compressed.</p></li>
</ol>
</div>
<div class="section" id="additional-arguments-for-the-fragments-function">
<h3>Additional arguments for the fragments function<a class="headerlink" href="#additional-arguments-for-the-fragments-function" title="Permalink to this headline">¶</a></h3>
<div class="section" id="number-of-processors-nproc">
<h4>Number of processors: <code class="docutils literal notranslate"><span class="pre">--nproc</span></code><a class="headerlink" href="#number-of-processors-nproc" title="Permalink to this headline">¶</a></h4>
<p>Multiple cores can be used by specifying the <code class="docutils literal notranslate"><span class="pre">--nproc</span></code> argument.
Specifying multiple processors will parallelize across chromosomes. Currently,
at most one thread is used per chromosome, so there is no point specifiying
more processors than the number of chromosomes.</p>
</div>
<div class="section" id="minimum-mapping-quality-min-mapq">
<h4>Minimum mapping quality: <code class="docutils literal notranslate"><span class="pre">--min_mapq</span></code><a class="headerlink" href="#minimum-mapping-quality-min-mapq" title="Permalink to this headline">¶</a></h4>
<p>The minimum allowed mapping quality (MAPQ) can be set using <code class="docutils literal notranslate"><span class="pre">--min_mapq</span></code>.
Depending on the aligner used, the MAPQ value can mean different <a class="reference external" href="https://sequencing.qcfail.com/articles/mapq-values-are-really-useful-but-their-implementation-is-a-mess/">things</a>.
Cellranger-atac uses <code class="docutils literal notranslate"><span class="pre">bwa-mem</span></code> for alignment, which follows the SAM spec and
reports Phred scores as MAPQ values:</p>
<blockquote>
<div><p>MAPping Quality.  It equals -10 log10 Pr {mapping position is wrong},
rounded to the nearest integer.  A value 255 indicates that the
mapping quality is not available.</p>
</div></blockquote>
</div>
<div class="section" id="cell-barcode-tag-barcodetag">
<h4>Cell barcode tag: <code class="docutils literal notranslate"><span class="pre">--barcodetag</span></code><a class="headerlink" href="#cell-barcode-tag-barcodetag" title="Permalink to this headline">¶</a></h4>
<p>Different methods may use different tags to store the cell barcode.
Cellranger uses the <code class="docutils literal notranslate"><span class="pre">CB</span></code> tag, which is set as the default for Sinto.
Other methods may use different tags, for example SNARE-seq uses the <code class="docutils literal notranslate"><span class="pre">XC</span></code> tag.
You can work out what tag is used by looking at part of the BAM file:
<code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">view</span> <span class="pre">aln.bam</span> <span class="pre">|</span> <span class="pre">head</span></code>.</p>
</div>
<div class="section" id="cell-barcode-regex-barcode-regex">
<h4>Cell barcode regex: <code class="docutils literal notranslate"><span class="pre">--barcode_regex</span></code><a class="headerlink" href="#cell-barcode-regex-barcode-regex" title="Permalink to this headline">¶</a></h4>
<p>Some methods store the cell barcode in the read name rather than under a read tag.
If this is the case, you can use a regular expression to extract the cell barcode
from the read name. For example, if the first section of your read name
up until the first <code class="docutils literal notranslate"><span class="pre">:</span></code> character corresponds to the cell barcode sequence,
you can specify <code class="docutils literal notranslate"><span class="pre">--barcode_regex</span> <span class="pre">[^:]*</span></code> to correcly match the cell barcodes.</p>
</div>
<div class="section" id="choosing-chromosomes-to-include-use-chrom">
<h4>Choosing chromosomes to include: <code class="docutils literal notranslate"><span class="pre">--use_chrom</span></code><a class="headerlink" href="#choosing-chromosomes-to-include-use-chrom" title="Permalink to this headline">¶</a></h4>
<p>Often a genome build might contain several scaffolds that are not typically used in
downstream analysis. This option allows you to specify a regular expression to
match chromosome names that will be retained in the output. By default,
all chromosomes starting with “chr” are retained, case insensitive (ie,
“Chr”, and “CHR” are also retained).</p>
</div>
<div class="section" id="set-the-maximum-distance-between-tn5-integration-sites-max-distance">
<h4>Set the maximum distance between Tn5 integration sites: <code class="docutils literal notranslate"><span class="pre">--max_distance</span></code><a class="headerlink" href="#set-the-maximum-distance-between-tn5-integration-sites-max-distance" title="Permalink to this headline">¶</a></h4>
<p>Incorrect alignment can sometimes generate implausible fragment coordinates.
Since we known there is an upper limit to the size of a DNA molecule that
can be sequenced on the Illumina platform, very large fragments over 5 kb
in size likely originate from incorrect read mapping. We can remove these
to reduce the impact of mapping artefacts on the downstream analysis
by setting the <code class="docutils literal notranslate"><span class="pre">--max_distance</span></code> parameter. Fragments larger than
this value will not be included in the output file.</p>
</div>
<div class="section" id="set-the-maximum-number-of-fragments-to-hold-in-memory-before-collapsing-chunksize">
<h4>Set the maximum number of fragments to hold in memory before collapsing: <code class="docutils literal notranslate"><span class="pre">--chunksize</span></code><a class="headerlink" href="#set-the-maximum-number-of-fragments-to-hold-in-memory-before-collapsing-chunksize" title="Permalink to this headline">¶</a></h4>
<p>The fragments algorithm iterates through a position-sorted BAM file and stores
fragment information as it iterates through the paired reads. Once all the
reads at a genomic locus have been read, the fragments covering that locus can
be PCR-collapsed. Sinto performs this step in chunks to balance speed and memory
use. The <code class="docutils literal notranslate"><span class="pre">--chunksize</span></code> parameter controls how many fragments are able to be
held in memory before they get collapsed and written to a file. Setting a larger
value should require more memory but the function will complete faster.</p>
</div>
</div>
</div>
<div class="section" id="filter-cell-barcodes-from-bam-file">
<h2>Filter cell barcodes from BAM file<a class="headerlink" href="#filter-cell-barcodes-from-bam-file" title="Permalink to this headline">¶</a></h2>
<p>Reads for a subset of cells can be extracted from a BAM file using the <code class="docutils literal notranslate"><span class="pre">filterbarcodes</span></code> command.
This requires a position-sorted, indexed BAM file, and a file containing a list of cell barcodes to retain.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> sinto filterbarcodes [-h] -b BAM -c CELLS -o OUTPUT [-t] [-s]
                         [-p NPROC] [--barcode_regex BARCODE_REGEX]
                         [--barcodetag BARCODETAG]

Filter reads based on input list of cell barcodes

optional arguments:
-h, --help            show this help message and exit
-b BAM, --bam BAM     Input bam file (must be indexed)
-c CELLS, --cells CELLS
                        File or comma-separated list of cell barcodes. Can be
                        gzip compressed
-t, --trim_suffix     Remove trail 2 characters from cell barcode in BAM
                        file
-p NPROC, --nproc NPROC
                        Number of processors (default = 1)
--barcode_regex BARCODE_REGEX
                        Regular expression used to extract cell barcode from
                        read name. If None (default), extract cell barcode
                        from read tag. Use &quot;[^:]*&quot; to match all characters up
                        to the first colon.
--barcodetag BARCODETAG
                        Read tag storing cell barcode information (default =
                        &quot;CB&quot;)
</pre></div>
</div>
<p>The input “cells” file should be a tab-delimited text file with cell barcodes in
the first column and the groups the cell belongs to in the second column. This
could be the cluster number, for example. A cell can belong to multiple groups
specified in the file using a comma-separated list of groups. If multiple
groups are provided, reads from that cell will be copied to the output BAM
file for each of the groups.</p>
<p>Example input “cells” file:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>TGGCAATGTTGAAGCG-1  A
GACCAATCACCATTCC-1  A
CAGGATTCAGAACTTC-1  B
GAACCTAAGAGAGGTA-1  B,A
ACATGGTGTAGACGCA-1  C
CCCTGATTCGGATAGG-1  C
</pre></div>
</div>
<p>The names of the output BAM files are determined by the name of each group in the
input cells file. The example file above would generate three bam files,
named <code class="docutils literal notranslate"><span class="pre">A.bam</span></code>, <code class="docutils literal notranslate"><span class="pre">B.bam</span></code>, and <code class="docutils literal notranslate"><span class="pre">C.bam</span></code>. Note that reads from the fourth cell
would appear in both <code class="docutils literal notranslate"><span class="pre">B.bam</span></code> and <code class="docutils literal notranslate"><span class="pre">A.bam</span></code>.</p>
</div>
<div class="section" id="add-read-tags-to-bam-file">
<h2>Add read tags to BAM file<a class="headerlink" href="#add-read-tags-to-bam-file" title="Permalink to this headline">¶</a></h2>
<p>Read tags can be added to a BAM file according to which cell the read belongs to using the <code class="docutils literal notranslate"><span class="pre">addtags</span></code> command.
This requires a position-sorted and indexed BAM file, and a file specifying the tags to be added to each cell, for example:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>TGGCAATGTTGAAGCG-1  CI      A
GACCAATCACCATTCC-1  CI      A
CAGGATTCAGAACTTC-1  CI      B
GAACCTAAGAGAGGTA-1  CI      B
ACATGGTGTAGACGCA-1  CI      C
CCCTGATTCGGATAGG-1  CI      C
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>sinto addtags [-h] -b BAM -f TAGFILE -o OUTPUT [-t] [-s] [-p NPROC]
                    [-m MODE]

Add read tags to reads from individual cells

optional arguments:
-h, --help            show this help message and exit
-b BAM, --bam BAM     Input bam file (must be indexed)
-f TAGFILE, --tagfile TAGFILE
                        Tab-delimited file containing cell barcode, tag to be
                        added, and tag identity. Can be gzip compressed
-o OUTPUT, --output OUTPUT
                        Name for output BAM file
-t, --trim_suffix     Remove trail 2 characters from cell barcode in BAM
                        file
-s, --sam             Output sam format (default bam output)
-p NPROC, --nproc NPROC
                        Number of processors (default = 1)
-m MODE, --mode MODE  Either tag (default) or readname. Some BAM file store
                        the cell barcode in the readname rather than under a
                        read tag
</pre></div>
</div>
<p>This will add a <code class="docutils literal notranslate"><span class="pre">CI</span></code> tag, with the tag set to A, B, or C depending on the cell barcode sequence.</p>
</div>
<div class="section" id="convert-read-tag-to-read-group">
<h2>Convert read tag to read group<a class="headerlink" href="#convert-read-tag-to-read-group" title="Permalink to this headline">¶</a></h2>
<p>Read groups can be added to a SAM/BAM file based on an arbitrary read tag using the
<code class="docutils literal notranslate"><span class="pre">tagtorg</span></code> command. Let’s assume we have a SAM file called <code class="docutils literal notranslate"><span class="pre">input.sam</span></code>
with the following contents:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@HD VN:1.5  SO:coordinate
@SQ SN:20   LN:63025520
@RG ID:rg1  SM:sample_1     LB:1    PU:1    PL:ILLUMINA
r002        0       20      9       30      3S6M1P1I4M      *       0       0       AAAAGATAAGGATA  *       CB:Z:AAAA-1     RG:Z:rg1
r003        0       20      9       30      3S6M1P1I4M      *       0       0       AAAAGATAAGGATA  *       CB:Z:CCCC-1     RG:Z:rg1
</pre></div>
</div>
<p>We would like to assign each read to a separate read group according to the value of
it’s <code class="docutils literal notranslate"><span class="pre">CB</span></code> tag. First, we need a list of tag values that we expect to see:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>AAAA-1
CCCC-1
</pre></div>
</div>
<p>Let us assume that the barcodes are stored in a file called <code class="docutils literal notranslate"><span class="pre">barcodes.txt</span></code>.
Then we can replace the read groups in the SAM file using the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sinto</span> <span class="n">tagtorg</span> <span class="o">-</span><span class="n">b</span> <span class="nb">input</span><span class="o">.</span><span class="n">sam</span> <span class="o">-</span><span class="n">f</span> <span class="n">barcodes</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>This will print the following SAM file to screen:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@HD</span> <span class="n">VN</span><span class="p">:</span><span class="mf">1.5</span>  <span class="n">SO</span><span class="p">:</span><span class="n">coordinate</span>
<span class="nd">@SQ</span> <span class="n">SN</span><span class="p">:</span><span class="mi">20</span>   <span class="n">LN</span><span class="p">:</span><span class="mi">63025520</span>
<span class="nd">@RG</span> <span class="n">ID</span><span class="p">:</span><span class="n">rg1</span>  <span class="n">SM</span><span class="p">:</span><span class="n">sample_1</span>     <span class="n">LB</span><span class="p">:</span><span class="mi">1</span>    <span class="n">PU</span><span class="p">:</span><span class="mi">1</span>    <span class="n">PL</span><span class="p">:</span><span class="n">ILLUMINA</span>
<span class="nd">@RG</span> <span class="n">ID</span><span class="p">:</span><span class="n">rg1</span><span class="p">:</span><span class="n">CCCC</span><span class="o">-</span><span class="mi">1</span>   <span class="n">SM</span><span class="p">:</span><span class="n">sample_1</span><span class="p">:</span><span class="n">CCCC</span><span class="o">-</span><span class="mi">1</span>      <span class="n">LB</span><span class="p">:</span><span class="mi">1</span>    <span class="n">PU</span><span class="p">:</span><span class="mi">1</span>    <span class="n">PL</span><span class="p">:</span><span class="n">ILLUMINA</span>
<span class="nd">@RG</span> <span class="n">ID</span><span class="p">:</span><span class="n">rg1</span><span class="p">:</span><span class="n">AAAA</span><span class="o">-</span><span class="mi">1</span>   <span class="n">SM</span><span class="p">:</span><span class="n">sample_1</span><span class="p">:</span><span class="n">AAAA</span><span class="o">-</span><span class="mi">1</span>      <span class="n">LB</span><span class="p">:</span><span class="mi">1</span>    <span class="n">PU</span><span class="p">:</span><span class="mi">1</span>    <span class="n">PL</span><span class="p">:</span><span class="n">ILLUMINA</span>
<span class="n">r002</span>        <span class="mi">0</span>       <span class="mi">20</span>      <span class="mi">9</span>       <span class="mi">30</span>      <span class="mi">3</span><span class="n">S6M1P1I4M</span>      <span class="o">*</span>       <span class="mi">0</span>       <span class="mi">0</span>       <span class="n">AAAAGATAAGGATA</span>  <span class="o">*</span>       <span class="n">CB</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="n">AAAA</span><span class="o">-</span><span class="mi">1</span>     <span class="n">RG</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="n">rg1</span><span class="p">:</span><span class="n">AAAA</span><span class="o">-</span><span class="mi">1</span>
<span class="n">r003</span>        <span class="mi">0</span>       <span class="mi">20</span>      <span class="mi">9</span>       <span class="mi">30</span>      <span class="mi">3</span><span class="n">S6M1P1I4M</span>      <span class="o">*</span>       <span class="mi">0</span>       <span class="mi">0</span>       <span class="n">AAAAGATAAGGATA</span>  <span class="o">*</span>       <span class="n">CB</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="n">CCCC</span><span class="o">-</span><span class="mi">1</span>     <span class="n">RG</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="n">rg1</span><span class="p">:</span><span class="n">CCCC</span><span class="o">-</span><span class="mi">1</span>
</pre></div>
</div>
<p>Two new &#64;RG tags have been added to the header with SM fields that are cell
barcode-specic. The two reads r002 and r003 have been assigned new RG tags
according to their cell barcode.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>usage: sinto tagtorg [-h] -b BAM [--tag TAG] -f TAGFILE [-o OUTPUT] [-O O]

Append a read tag to the read group ID of each read. Also appends the read tag
to the SM field of the read group.

optional arguments:
  -h, --help            show this help message and exit
  -b BAM, --bam BAM     Input SAM/BAM file, &#39;-&#39; reads from stdin
  --tag TAG             Read tag to extract the value from that is appended to
                        the read group. Default is &#39;CB&#39;, the tag that is used
                        in 10x sequencing to identify cells.
  -f TAGFILE, --tagfile TAGFILE
                        List of expected tag values. Reads with tag values
                        that are not in this list are not altered.
  -o OUTPUT, --output OUTPUT
                        Output SAM/BAM file, &#39;-&#39; outputs to stdout (default
                        &#39;-&#39;)
  -O OUTPUTFORMAT, --outputformat OUTPUTFORMAT
                        Output format. One of &#39;t&#39; (SAM), &#39;b&#39; (BAM), or &#39;u&#39;
                        (uncompressed BAM) (&#39;t&#39; default)
</pre></div>
</div>
</div>
<div class="section" id="copy-move-read-tag-to-another-read-tag">
<h2>Copy/move read tag to another read tag<a class="headerlink" href="#copy-move-read-tag-to-another-read-tag" title="Permalink to this headline">¶</a></h2>
<p>Read tags can be renamed or copied to anthor read tag using the <code class="docutils literal notranslate"><span class="pre">tagtotag</span></code> command.
Let’s assume we have a SAM file called <code class="docutils literal notranslate"><span class="pre">input.sam</span></code>
with the following contents:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@HD VN:1.5  SO:coordinate
@SQ SN:20   LN:63025520
r002        0       20      9       30      3S6M1P1I4M      *       0       0       AAAAGATAAGGATA  *       CB:Z:AAAA-1
r003        0       20      9       30      3S6M1P1I4M      *       0       0       AAAAGATAAGGATA  *       CB:Z:CCCC-1
</pre></div>
</div>
<p>We would like to rename the CB tag to another arbitrary tag, let’s call it xx.
If we run the following command:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>sinto tagtotag --from CB --to xx --delete --bam - -o -
</pre></div>
</div>
<p>This will print the following SAM file to screen:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@HD VN:1.5  SO:coordinate
@SQ SN:20   LN:63025520
r002        0       20      9       30      3S6M1P1I4M      *       0       0       AAAAGATAAGGATA  *       xx:Z:AAAA-1
r003        0       20      9       30      3S6M1P1I4M      *       0       0       AAAAGATAAGGATA  *       xx:Z:CCCC-1
</pre></div>
</div>
<p>The two CB tags have been renamed to xx. If we wish to keep the original CB tag, then
we can drop <code class="docutils literal notranslate"><span class="pre">--delete</span></code> from the command.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>usage: sinto tagtotag [-h] -b BAM --from FROM_ --to TO [--delete] [-o OUTPUT]
                      [-O OUTPUTFORMAT]

Copies BAM entries to a new file while copying a read tag to another read tag
and optionally deleting the originating tag.

optional arguments:
  -h, --help            show this help message and exit
  -b BAM, --bam BAM     Input SAM/BAM file, &#39;-&#39; reads from stdin
  --from FROM_          Read tag to copy from.
  --to TO               Read tag to copy to.
  --delete              Delete originating tag after copy (i.e. move).
  -o OUTPUT, --output OUTPUT
                        Output SAM/BAM file, &#39;-&#39; outputs to stdout (default
                        &#39;-&#39;)
  -O OUTPUTFORMAT, --outputformat OUTPUTFORMAT
                        Output format. One of &#39;t&#39; (SAM), &#39;b&#39; (BAM), or &#39;u&#39;
                        (uncompressed BAM) (&#39;t&#39; default)
</pre></div>
</div>
</div>
<div class="section" id="add-cell-barcodes-to-fastq-read-names">
<h2>Add cell barcodes to FASTQ read names<a class="headerlink" href="#add-cell-barcodes-to-fastq-read-names" title="Permalink to this headline">¶</a></h2>
<p>Cell barcodes from one FASTQ file can be added to the read names of another, or the same,
FASTQ file using the <code class="docutils literal notranslate"><span class="pre">barcode</span></code> command. This is useful when processing raw single-cell
sequencing data, as the cell barcode information can easily be propagated to the aligned
BAM file by encoding the cell barcode in the read name. Both gzipped and uncompressed
FASTQ files are supported as input. Running on uncompressed FASTQ is usually much faster
than running on gzipped FASTQ files.</p>
<p>Running this command will generate new gzipped FASTQ files with the read names modified to
contain the cell barcode sequence at the beginning of the read name, separated from the
original read name by a <code class="docutils literal notranslate"><span class="pre">:</span></code> character. The output files will be the name of the input
file with <code class="docutils literal notranslate"><span class="pre">.barcoded.fastq.gz</span></code> at the end of the file name.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>sinto barcode [-h] --barcode_fastq BARCODE_FASTQ --read1 READ1
                 [--read2 READ2] -b BASES [--prefix PREFIX]
                 [--suffix SUFFIX]

Add cell barcode sequences to read names in FASTQ file.

optional arguments:
-h, --help            show this help message and exit
--barcode_fastq BARCODE_FASTQ
                        FASTQ file containing cell barcode sequences
--read1 READ1         FASTQ file containing read 1
--read2 READ2         FASTQ file containing read 2
-b BASES, --bases BASES
                        Number of bases to extract from barcode-containing
                        FASTQ
--prefix PREFIX       Prefix to add to cell barcodes
--suffix SUFFIX       Suffix to add to cell barcodes
</pre></div>
</div>
<div class="section" id="additional-arguments-for-the-barcode-function">
<h3>Additional arguments for the barcode function<a class="headerlink" href="#additional-arguments-for-the-barcode-function" title="Permalink to this headline">¶</a></h3>
<div class="section" id="bases-bases">
<h4>Bases: <code class="docutils literal notranslate"><span class="pre">--bases</span></code><a class="headerlink" href="#bases-bases" title="Permalink to this headline">¶</a></h4>
<p>This controls how many bases from the read containing the cell barcode are used.
Bases are counted from the beginning of the read sequence in the FASTQ file. For
example, <code class="docutils literal notranslate"><span class="pre">--bases</span> <span class="pre">12</span></code> will extract the first 12 sequenced bases from the read
and use it as the cell barcode.</p>
</div>
<div class="section" id="barcode-read-file-barcode-fastq">
<h4>Barcode read file: <code class="docutils literal notranslate"><span class="pre">--barcode_fastq</span></code><a class="headerlink" href="#barcode-read-file-barcode-fastq" title="Permalink to this headline">¶</a></h4>
<p>FASTQ file with reads containing the cell barcode sequence.</p>
</div>
<div class="section" id="read-1-and-read-2-read1-and-read2">
<h4>Read 1 and read 2: <code class="docutils literal notranslate"><span class="pre">--read1</span></code> and <code class="docutils literal notranslate"><span class="pre">--read2</span></code><a class="headerlink" href="#read-1-and-read-2-read1-and-read2" title="Permalink to this headline">¶</a></h4>
<p>FASTQ files containing reads to which the cell barcode information will be
added. Note that these files must contain the same number of reads as the barcode-containing
FASTQ file, and the reads must appear in the same order.</p>
</div>
</div>
<div class="section" id="example">
<h3>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h3>
<p>Take the following two FASTQ files as an example. The first contains cell barcode sequences
and the second we want to add those sequences to the read name.</p>
<p><code class="docutils literal notranslate"><span class="pre">barocde_file.fastq.gz</span></code>:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@D00611:697:CD0V6ANXX:5:2301:1176:2478 1:N:0:TATCCTCT
CAATACACTATATGGGAGACGTTTTTTTTT
+
BBBBBFFFFFFFFFFFFFFFFFFFFFFFFF
@D00611:697:CD0V6ANXX:5:2301:1480:2408 1:N:0:TATCCTCT
CAGAGACGTAAACAATGGCGGTTTTTTTTT
+
B&lt;BBBFFFFFFFFFFFFFFFFFFFFFFFFF
@D00611:697:CD0V6ANXX:5:2301:1361:2447 1:N:0:TATCCTAT
AGTCTCGCCACATGGGGGGGATTTTTTTTT
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">read1.fastq.gz</span></code>:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@D00611:697:CD0V6ANXX:5:2301:1176:2478 2:N:0:TATCCTCT
GATTTACACAGATGATATGTTTCTATTGCCTGCTTGGGATGGGGGTGGGAGGCAGAGTCCATCTACCTCTCTAAC
+
BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@D00611:697:CD0V6ANXX:5:2301:1480:2408 2:N:0:TATCCTCT
GTGCCTTTGACTTTAGCTAGGCGACAGGGGACGAGTCCATTAGCATACNNNGTAAATTGCTGTTGTCTGTTTTTG
+
&lt;////////B/B/////&lt;//////&lt;///////////&lt;///////////###////////////&lt;///////////
@D00611:697:CD0V6ANXX:5:2301:1361:2447 2:N:0:TATCCTAT
TAATACATGACGGTGTCTTAGTAGCACTTACTATGCACAGGTTAAGACCTGTCTCTTATACACATCTCCGAGCCC
</pre></div>
</div>
<p>After running <code class="docutils literal notranslate"><span class="pre">sinto</span> <span class="pre">barcode</span></code> with <code class="docutils literal notranslate"><span class="pre">-b</span> <span class="pre">12</span></code> to extract the first 12 bases of the barcode sequence
we have a new file called <code class="docutils literal notranslate"><span class="pre">read1.barcoded.fastq.gz</span></code>:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@CAATACACTATA:D00611:697:CD0V6ANXX:5:2301:1176:2478 2:N:0:TATCCTCT
GATTTACACAGATGATATGTTTCTATTGCCTGCTTGGGATGGGGGTGGGAGGCAGAGTCCATCTACCTCTCTAAC
+
BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@CAGAGACGTAAA:D00611:697:CD0V6ANXX:5:2301:1480:2408 2:N:0:TATCCTCT
GTGCCTTTGACTTTAGCTAGGCGACAGGGGACGAGTCCATTAGCATACNNNGTAAATTGCTGTTGTCTGTTTTTG
+
&lt;////////B/B/////&lt;//////&lt;///////////&lt;///////////###////////////&lt;///////////
@AGTCTCGCCACA:D00611:697:CD0V6ANXX:5:2301:1361:2447 2:N:0:TATCCTAT
TAATACATGACGGTGTCTTAGTAGCACTTACTATGCACAGGTTAAGACCTGTCTCTTATACACATCTCCGAGCCC
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="scatac.html" class="btn btn-neutral float-right" title="Processing scATAC-seq data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Tim Stuart, Warren W. Kretzschmar

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>