

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Processing scATAC-seq data &mdash; sinto 0.7.6 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="User guide" href="basic_usage.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> sinto
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="basic_usage.html">User guide</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Processing scATAC-seq data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#download-data">Download data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#attach-cell-barcodes">Attach cell barcodes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#map">Map</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-a-fragment-file">Create a fragment file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#call-peaks">Call peaks</a></li>
<li class="toctree-l2"><a class="reference internal" href="#downstream-analysis">Downstream analysis</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sinto</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Processing scATAC-seq data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/scatac.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="processing-scatac-seq-data">
<h1>Processing scATAC-seq data<a class="headerlink" href="#processing-scatac-seq-data" title="Permalink to this headline">¶</a></h1>
<p>In this tutorial we’ll demonstrate a simple pipeline for processing scATAC-seq data using Sinto,
<a class="reference external" href="https://github.com/lh3/bwa">bwa</a>, <a class="reference external" href="https://github.com/samtools/htslib">tabix</a>, and <a class="reference external" href="https://github.com/jsh58/Genrich">Genrich</a>.</p>
<div class="section" id="download-data">
<h2>Download data<a class="headerlink" href="#download-data" title="Permalink to this headline">¶</a></h2>
<p>For this example we’ll download scATAC-seq data from
<a class="reference external" href="https://www.nature.com/articles/s41587-019-0290-0">Chen, Lake, and Zhang (2019)</a>
This dataset actually contains both gene expression and DNA accessibility measurements for each
cell, but we can process the DNA accessibility data exactly as though it was a typical scATAC-seq
experiment. In this example we’ll just process one replicate, but there are 12 replicates in
total in the original data.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget https://sra-pub-src-1.s3.amazonaws.com/SRR9672090/SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R1_001.fastq.gz.1
wget https://sra-pub-src-1.s3.amazonaws.com/SRR9672090/SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R2_001.fastq.gz.1
wget https://sra-pub-src-1.s3.amazonaws.com/SRR9672090/SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R3_001.fastq.gz.1
</pre></div>
</div>
</div>
<div class="section" id="attach-cell-barcodes">
<h2>Attach cell barcodes<a class="headerlink" href="#attach-cell-barcodes" title="Permalink to this headline">¶</a></h2>
<p>In this experiment the cell barcodes were sequenced using the first 12 bp of read 1,
and the paired-end genomic DNA reads sequenced using read 2 and read 3. We can use the
<code class="docutils literal notranslate"><span class="pre">sinto</span> <span class="pre">barcode</span></code> function to transfer the cell barcodes from read 1 into the read names
in read 2 and read 3. Storing the cell barcode in the read name is an easy way to track
which reads came from which cells.</p>
<p>This step is much faster when using unzipped fastq files, so we’ll first unzip the fastq
files we downloaded earlier.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gzip -S .gz.1 -d SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R1_001.fastq.gz.1
gzip -S .gz.1 -d SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R2_001.fastq.gz.1
gzip -S .gz.1 -d SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R3_001.fastq.gz.1
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sinto barcode -b <span class="m">12</span> --barcode_fastq SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R1_001.fastq <span class="se">\</span>
    --read1 SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R2_001.fastq <span class="se">\</span>
    --read2 SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R3_001.fastq
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Function run_barcode called with the following arguments:

barcode_fastq       SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R1_001.fastq
read1       SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R2_001.fastq
read2       SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R3_001.fastq
bases       12
prefix
suffix
func        &lt;function run_barcode at 0x7f28c93714c0&gt;

Function completed in  8.0 m 13.35 s
</pre></div>
</div>
</div>
<div class="section" id="map">
<h2>Map<a class="headerlink" href="#map" title="Permalink to this headline">¶</a></h2>
<p>We can map the paired-end genomic reads using any aligner we like. In this example we’ll use
bwa mem. First we need to download the fasta file for the mm10 genome and create a genome index for bwa:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget https://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/mm10.fa.gz
gzip -d mm10.fa.gz
bwa index mm10.fa
</pre></div>
</div>
<p>Next we can map the reads with bwa mem:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bwa mem -t <span class="m">8</span> mm10.fa SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R2_001.barcoded.fastq <span class="se">\</span>
    SC_20190509A_AdMsBrain_SNARE_chromatin_S16_R3_001.barcoded.fastq <span class="se">\</span>
    <span class="p">|</span> samtools view -b - &gt; aln.bam

<span class="c1"># sort and index bam file</span>
samtools sort -@ <span class="m">8</span> aln.bam -o aln.sort.bam
samtools index -@ <span class="m">8</span> aln.sort.bam
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># look at some of the aligned reads</span>
samtools view aln.sort.bam <span class="p">|</span> head -n <span class="m">3</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>AACCTCCAACTG:D00611:698:CDN4HANXX:5:1106:1580:94969 69      chr1    3000052 0       *       =       3000052 0       TCACATTCTCAGTGCACAATAGAACCCCTTACCTCCAATCCAGAGTAAACAAAGAGTCTACCACAAACACACATC     /&lt;/&lt;///&lt;/BB/////&lt;//////&lt;///&lt;//&lt;&lt;////&lt;B&lt;BBB///////&lt;&lt;//////&lt;&lt;//&lt;////&lt;/&lt;&lt;FF&lt;&lt;/     MC:Z:32M43S     AS:i:0  XS:i:0
AACCTCCAACTG:D00611:698:CDN4HANXX:5:1106:1580:94969 137     chr1    3000052 0       32M43S  =       3000052 0       TTGAAGGTCTGGTAGAACTCTGCATTAAACCCCGACTCCAGTTGGAGGTTGTACTCTGCGTTGATACCACTTTTT     BBB/BFFFBFBFFFFFBFFFFFFFFFFFFFFFFFFFFFFFFFFFFBFFFFFFFFFFFFFFFFFFFFB&lt;///////     NM:i:0  MD:Z:32 AS:i:32 XS:i:32
TCCGCCGGAAAC:D00611:697:CD0V6ANXX:7:2116:7385:7732  117     chr1    3000098 0       *       =       3000098 0       AAGACGGCATACGAGATTCGCCTTAGTCTCGTGGGCTCGGAGATGTGTATAAGAGACATATACACTCAGCTTTAT     FFFFFFFFFFFFFFFFFFFFFFFFFFFF&lt;FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB     MC:Z:33M42S     AS:i:0  XS:i:0
</pre></div>
</div>
</div>
<div class="section" id="create-a-fragment-file">
<h2>Create a fragment file<a class="headerlink" href="#create-a-fragment-file" title="Permalink to this headline">¶</a></h2>
<p>Finally we create a fragment file for the dataset using <code class="docutils literal notranslate"><span class="pre">sinto</span> <span class="pre">fragments</span></code></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sinto fragments -b aln.sort.bam -p <span class="m">8</span> -f fragments.bed --barcode_regex <span class="s2">&quot;[^:]*&quot;</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Function run_fragments called with the following arguments:

bam aln.sort.bam
fragments   fragments.bed
min_mapq    30
nproc       8
barcodetag  CB
cells       None
barcode_regex       [^:]*
use_chrom   (?i)^chr
max_distance        5000
chunksize   500000
func        &lt;function run_fragments at 0x7ff6976b5dc0&gt;

Function completed in  7.0 m 30.24 s
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort, compress, and index</span>
sort -k1,1 -k2,2n fragments.bed &gt; fragments.sort.bed
bgzip -@ <span class="m">8</span> fragments.sort.bed
tabix -p bed fragments.sort.bed.gz

<span class="c1"># clean up</span>
rm fragments.bed

<span class="c1"># take a look at the output</span>
gzip -dc fragments.sort.bed.gz <span class="p">|</span> head
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chr1        3003930 3003960 ATGGTGACTCAT    1
chr1        3010068 3010136 CTCGGGTTTAGC    5
chr1        3010538 3010589 GAAACGCAAAGT    8
chr1        3011663 3011770 AACTGGGCTATC    2
chr1        3011684 3011728 CTATTGTGCATA    12
chr1        3012702 3012739 ATCAGAGACGGC    6
chr1        3012724 3012775 TGCGGACGGTGG    3
chr1        3012743 3012804 TGACATCCCCGA    2
chr1        3019505 3019588 CACGCTCGTCTT    1
chr1        3021571 3021620 GTATTCGTCGGG    9
</pre></div>
</div>
</div>
<div class="section" id="call-peaks">
<h2>Call peaks<a class="headerlink" href="#call-peaks" title="Permalink to this headline">¶</a></h2>
<p>We can call peaks using all cells combined using the mapped bam file and <a class="reference external" href="https://github.com/jsh58/Genrich">Genrich</a>.
Genrich requires reads to be sorted by queryname, so we first re-sort the bam file and
then run Genrich using ATAC-seq mode (<code class="docutils literal notranslate"><span class="pre">-j</span></code>) to find peaks.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>samtools sort -n -@ <span class="m">8</span> aln.sort.bam -o aln.qname.bam
Genrich -j -t aln.qname.bam -o peaks.bed -v
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># look at some of the peaks</span>
head peaks.bed
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chr1        3094859 3095377 peak_0  1000    .       913.693115      5.720295        -1      374
chr1        3119581 3120170 peak_1  1000    .       1186.086792     6.389130        -1      420
chr1        3120272 3120746 peak_2  1000    .       965.524841      6.166295        -1      285
chr1        3121354 3121654 peak_3  1000    .       580.268433      4.948659        -1      129
chr1        3292563 3292985 peak_4  1000    .       897.739197      5.319649        -1      257
chr1        3299692 3299934 peak_5  1000    .       489.207764      5.154057        -1      98
chr1        3309998 3310407 peak_6  1000    .       829.686340      6.153750        -1      210
chr1        3322437 3322775 peak_7  1000    .       397.093872      3.806610        -1      128
chr1        3360973 3361167 peak_8  1000    .       262.217590      3.911148        -1      120
chr1        3369527 3369829 peak_9  1000    .       373.499207      4.094081        -1      149
</pre></div>
</div>
</div>
<div class="section" id="downstream-analysis">
<h2>Downstream analysis<a class="headerlink" href="#downstream-analysis" title="Permalink to this headline">¶</a></h2>
<p>Downstream analysis steps, including quantifying counts in each peak for each cell,
can be performed using <a class="reference external" href="https://satijalab.org/signac">Signac</a>.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="basic_usage.html" class="btn btn-neutral float-left" title="User guide" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, Tim Stuart, Warren W. Kretzschmar.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>